### PR TITLE
Add sandboxMode to all calls.

### PR DESCRIPTION
**Pull request by Alvtag**
**Description**
Add sandbox mode header to all calls

**Issues that should be CLOSED by merge of this PR:**
http://guestlogix.atlassian.net/jira/software/projects/TPP/boards/4?label=Android&selectedIssue=TPP-924

### CODE REVIEW

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/models/Traveler.java [hunk 1]
```diff
@@ -34,21 +33,21 @@
      * @param apiKey             key provided by the Guestlogix platform. Invalid keys will result in forbidden response codes.
      * @param applicationContext application context where the sdk is running.
      */
-    public static void initialize(String apiKey, Context applicationContext) {
+    public static void initialize(String apiKey, Context applicationContext, boolean sandboxMode) {
```
Comment (jack-mo) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/models/Traveler.java]: You're forcing the end users to use a sandbox mode boolean. Sandbox mode should be off by default. I think overloading the constructor here is will be better. Let me know what you think.

Comment (Alvtag) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/models/Traveler.java]: updated

**Comment by Alvtag** [comment]
> The boolean `sandboxMode` should be renamed to `isSandboxMode`

updated

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java [hunk 2]
```diff
@@ -204,14 +204,11 @@ public static AuthenticatedUrlRequest emailOrderConfirmation(Order order, Sessio
     private static class RouteBuilder {
         private NetworkTask.Route.Method method;
         private String path;
-        private String url;
         private JSONPayloadProvider payload = null;
-        private String apiKey;
         private Map<String, String> headers;
         private List<Pair<String, String>> params;
-        private Context context;
```
Comment (Alvtag) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: I removed three unused fields

Comment (ata-n) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: +1

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthenticatedRemoteNetworkRequestTask.java [hunk 3]
```diff
@@ -39,11 +41,12 @@ public void onError(Error error) {
         } else {
             this.responseHandler = null;
         }
+
+        authTokenFetchTask = new AuthTokenFetchTask(session, context);
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthenticatedRemoteNetworkRequestTask.java]: should create the task as part of the execute implementation.

Reason is
a) The work involve in allocating memory etc will be done the background thread
b) The object reference is not held on to for the entirely of this tasks existence

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java [hunk 4]
```diff
@@ -11,14 +12,17 @@
...
 
-    public AuthTokenFetchTask(String apiKey, Context context) {
-        this.context = context;
-        this.apiKey = apiKey;
+    private NetworkTask.Route route;
+    private SharedPrefsWriteTask sharedPrefsWriteTask;
+
+    public AuthTokenFetchTask(Session session, Context context) {
+
+        //Writing encrypted token to shared prefs
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java]: comments:
`// My comment`

notice the whitespace between slashes and comment

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java [hunk 5]
```diff
@@ -11,14 +12,17 @@
...
+        sharedPrefsWriteTask= new SharedPrefsWriteTask(context, session.getApiKey());
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java]: avoid instance variables holding onto tasks.

This creates memory leaks

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java [hunk 6]
```diff
@@ -1,6 +1,7 @@
 package com.guestlogix.travelercorekit.tasks;
 
 import android.content.Context;
+import com.guestlogix.travelercorekit.models.Session;
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java]: this class should not know of `Session`

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthenticatedRemoteNetworkRequestTask.java [hunk 7]
```diff
@@ -39,11 +41,12 @@ public void onError(Error error) {
...
     }
 
     @Override
     public void execute() {
-        AuthTokenFetchTask authTokenFetchTask = new AuthTokenFetchTask(this.session.getApiKey(), session.getContext());
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthenticatedRemoteNetworkRequestTask.java]: Now I see why `Session` included context

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java (hunk 5 above)
Comment (ata-n) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java]: code formatting ^

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java [hunk 8]
```diff
@@ -240,13 +237,13 @@ private RouteBuilder(Context context, String apiKey) {
             this.headers.put("x-region", region);
             this.headers.put("x-application-id", applicationId);
             this.headers.put("x-timezone", "UTC");
-            this.headers.put("x-api-key", apiKey);
-            this.context = context;
+            this.headers.put("x-api-key", session.getApiKey());
+
+            SharedPreferences sharedPreferences = context.getSharedPreferences("TRAVELER_PREFS", Context.MODE_PRIVATE);
+            travelerSDKEndpoint = sharedPreferences.getString("TRAVELER_SDK_ENDPOINT", DEFAULT_ENDPOINT);
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: We should not be reading from shared prefs everytime. This is super inefficient!

Comment (ata-n) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: I realize this was being done this way before as well. but we should fix this ASAP.

Hint: Use a lazy loaded static variable

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java [hunk 9]
```diff
@@ -178,42 +175,42 @@ public static AuthenticatedUrlRequest orders(OrderQuery query, Session session,
...
         private JSONPayloadProvider payload = null;
-        private String apiKey;
         private Map<String, String> headers;
         private List<Pair<String, String>> params;
-        private Context context;
+        private Session session;
+        private String travelerSDKEndpoint;
 
-        private RouteBuilder(Context context, String apiKey) {
+        private RouteBuilder(Context context, Session session) {
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: Im struggling to see the value of having this class have a reference to `Session` or even know of it.

Best if it doesn't and is model agnostic. Maybe I missed sth, but I don't see any use other than token here, which is best passed as an argument to the build method.

What happens now is that this class is making an assumption that session will store token and the token is valid for the duration of the session. While that may be true it is not within this classes responsibility scope to be aware of .

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthenticatedRemoteNetworkRequestTask.java (hunk 3 above)
Comment (Alvtag) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthenticatedRemoteNetworkRequestTask.java]: updated

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java (hunk 4 above)
Comment (Alvtag) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java]: cleaned up comments

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java (hunk 5 above)
Comment (Alvtag) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java]: updated

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java (hunk 6 above)
Comment (Alvtag) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/tasks/AuthTokenFetchTask.java]: updated

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java (hunk 9 above)
Comment (Alvtag) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: Given that if router doesn't have knowledge of session, session's fields will have to be passed in piecemeal as needed.

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java [hunk 10]
```diff
@@ -19,89 +19,87 @@
...
 import java.util.*;
 
-
 public class Router {
     private static final String DEFAULT_ENDPOINT = "https://traveler.rc.guestlogix.io/v1";
     private static String TAG = "Traveler";
 
-    public static UnauthenticatedUrlRequest authenticate(String apiKey, Context context) {
-        return new RouteBuilder(context, apiKey)
+    public static UnauthenticatedUrlRequest authenticate(Token token, String apiKey, Context context) {
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: Why does an unauthenticated request need a token?

File: traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java [hunk 11]
```diff
@@ -19,89 +19,87 @@
 import java.net.URL;
 import java.util.*;
 
-
 public class Router {
     private static final String DEFAULT_ENDPOINT = "https://traveler.rc.guestlogix.io/v1";
     private static String TAG = "Traveler";
```
Comment (ata-n) [inline traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: Maybe change this TAG as well while youre at it

Comment (Alvtag) [reply traveler_core_kit/src/main/java/com/guestlogix/travelercorekit/Router.java]: removed

**Pull request merged**